File: mobile/lib/features/doctorlist/data/models/doctor.dart
import 'package:mobile/features/doctorlist/domain/entities/doctorentity.dart';

class DoctorModel extends Doctor {
  const DoctorModel(
      {required super.doctorId,
      required super.categoryId,
      required super.name,
      required super.telephone,
      required super.email,
      required super.slmcNo,
      required super.birthday,
      required super.remarks,
      required super.frequency,
      required super.specialityId,
      required super.townId});

  factory DoctorModel.fromJson(Map<String, dynamic> json) {
    return DoctorModel(
      doctorId: json['doctor_id'],
      categoryId: json['category_id'],
      name: json['name'],
      telephone: json['telephone'],
      email: json['email'],
      slmcNo: json['slmc_no'],
      birthday: DateTime.parse(json['birthday']),
      remarks: json['remarks'],
      frequency: json['frequency'],
      specialityId: json['speciality_id'],
      townId: json['town_id'],
    );
  }
}

File: backend/models/definedModels/Doctor.js
// src/models/Doctor.js
const { Model, DataTypes } = require('sequelize');
const DoctorSpeciality = require('./DoctorSpeciality'); // Import the speciality model
const Town = require('./Towns'); // Import the town model
const DoctorCategory = require('./DoctorCategory'); // Import the doctor category model

class Doctor extends Model {
  // Static method to initialize the model with a dynamic sequelize instance
  static initModel(sequelize) {
    Doctor.init({
      doctor_id: {
        type: DataTypes.INTEGER,
        primaryKey: true,
        autoIncrement: true,
      },
      category_id: {
        type: DataTypes.INTEGER,
      },
      name: {
        type: DataTypes.STRING,
        allowNull: false,
      },
      telephone: {
        type: DataTypes.STRING,
      },
      email: {
        type: DataTypes.STRING,
      },
      slmc_no: {
        type: DataTypes.STRING,
      },
      birthday: {
        type: DataTypes.DATE,
      },
      remarks: {
        type: DataTypes.TEXT,
      },
      frequency: {
        type: DataTypes.INTEGER,
      },
      speciality_id: {
        type: DataTypes.INTEGER,
      },
      town_id: {
        type: DataTypes.INTEGER,
      },
    }, {
      sequelize, // Dynamically injected sequelize instance
      modelName: "Doctor", // Class name (can be pluralized if needed)
      tableName: 'doctors', // Explicit table name
      timestamps: false, // Disable automatic timestamp fields
    });

      // Doctor belongs to a DoctorCategory (many-to-one relationship)
      Doctor.belongsTo(DoctorCategory, {
        foreignKey: 'category_id',
        as: 'category',
      });
  
      // Doctor belongs to a DoctorSpeciality (many-to-one relationship)
      Doctor.belongsTo(DoctorSpeciality, {
        foreignKey: 'speciality_id',
        as: 'speciality',
      });
  
      // Doctor belongs to a Town (many-to-one relationship)
      Doctor.belongsTo(Town, {
        foreignKey: 'town_id',
        as: 'town',
      });
  }

  
}

module.exports = Doctor;